### PR TITLE
Feature/saa 464 add adjudications to scheduled events

### PR DESCRIPTION
**Pull request by matt-crowson-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt [hunk 1]
```diff
@@ -1,27 +1,11 @@
...
-import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.model.AppointmentLocationSummary
 import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.model.PayPerSession
-import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.model.PrisonerSummary
-import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.model.UserSummary
 import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.model.response.PrisonerAllocations
 import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.service.Priority
-import java.time.LocalDateTime
-import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.client.prisonapi.model.CourtHearings as PrisonApiCourtHearings
-import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.client.prisonapi.model.PrisonerSchedule as PrisonApiPrisonerSchedule
-import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.client.prisonapi.model.ScheduledEvent as PrisonApiScheduledEvent
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt:24]: TransformFunctions has been split out into 4 to represent the respective source.  This needs to be renamed as well (will rename after merge into main to reduce chances of merge conflicts).

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  prisonCode: String,
+  priorities: List<Priority>?,
+): List<ModelScheduleEvent> =
+  map { it.toScheduledEvent(prisonCode, priorities) }
+
+fun OffenderAdjudicationHearing.toScheduledEvent(
+  prisonCode: String,
+  priorities: List<Priority>?,
+) = let {
+  ScheduledEvent(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:115]: @timharrison-moj 

Not 100% on this transformation in terms on what we need and which fields some of the values have been applied to.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt (hunk 1 above)
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/TransformFunctions.kt:24]: The unit test for this probably needs breaking up as well.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/api/PrisonApiClient.kt [hunk 3]
```diff
@@ -273,7 +299,12 @@ class PrisonApiClient(private val prisonApiWebClient: WebClient) {
       .retrieve()
       .awaitBody()
 
-  // TODO: Replaced by async version below
+  @Deprecated(
+    message = "Replaced by coroutine version",
+    replaceWith = ReplaceWith(
+      expression = "getExternalTransfersOnDateAsync(agencyId, prisonerNumbers, date)",
+    ),
+  )
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/api/PrisonApiClient.kt:307]: Very good. I should have done that.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/model/OffenderAdjudicationHearing.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * @param hearingType Hearing Type
+ * @param startTime Hearing Time
+ * @param internalLocationDescription The internal location description of the hearing
+ * @param eventStatus The status of the hearing, SCH, COMP or EXP
+ */
+data class OffenderAdjudicationHearing(
+
+  @Schema(example = "null", required = true, description = "")
+  @get:JsonProperty("agencyId", required = true) val agencyId: kotlin.String,
+
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/model/OffenderAdjudicationHearing.kt:22]: Or are generated?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/model/OffenderAdjudicationHearing.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  @Schema(example = "789448", required = true, description = "The internal location id of the hearing")
+  @get:JsonProperty("internalLocationId", required = true) val internalLocationId: kotlin.Long,
+
+  @Schema(example = "Governor's Hearing Adult", description = "Hearing Type")
+  @get:JsonProperty("hearingType") val hearingType: kotlin.String? = null,
+
+  @get:Pattern(regexp = "^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}$")
+  @Schema(example = "2021-07-05T10:35:17", description = "Hearing Time")
+  @get:JsonProperty("startTime") val startTime: kotlin.String? = null,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/model/OffenderAdjudicationHearing.kt:37]: Will these need a `@JsonFormat()` to return the date and time in the right format?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/model/OffenderAdjudicationHearing.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/model/OffenderAdjudicationHearing.kt:22]: Generated.  I wonder if there is some setting with the generator?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/model/OffenderAdjudicationHearing.kt (hunk 5 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/client/prisonapi/model/OffenderAdjudicationHearing.kt:37]: Generated.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventService.kt [hunk 6]
```diff
@@ -95,12 +104,14 @@ class ScheduledEventService(
                 EventType.ACTIVITY.defaultPriority,
                 eventPriorities[EventType.ACTIVITY],
               ),
-              schedules.transfers.prisonApiPrisonerScheduleToScheduledEvents(
+              schedules.transfers.prisonApiTransfersToScheduledEvents(
                 prisonCode,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventService.kt:108]: Yay! You've started adding the type-specific transforms. Thanks - I had this on my list of things to do when time presents itself.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventService.kt [hunk 7]
```diff
@@ -223,36 +239,30 @@ class ScheduledEventService(
...
             eventPriorities[EventType.ACTIVITY],
           ),
-          schedules.transfers.prisonApiPrisonerScheduleToScheduledEvents(
+          schedules.transfers.prisonApiTransfersToScheduledEvents(
             prisonCode,
-            EventType.EXTERNAL_TRANSFER.name,
-            EventType.EXTERNAL_TRANSFER.defaultPriority,
             eventPriorities[EventType.EXTERNAL_TRANSFER],
           ),
+          schedules.adjudications.prisonApiOffenderAdjudicationsToScheduledEvents(
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ScheduledEventService.kt:262]: Like it.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    eventType = EventType.ADJUDICATION_HEARING.name,
+    eventTypeDesc = null,
+    event = null,
+    eventDesc = null,
+    details = null,
+    prisonerNumber = it.offenderNo,
+    date = LocalDateTime.parse(it.startTime).toLocalDate(),
+    startTime = LocalDateTime.parse(it.startTime).toLocalTime(),
+    endTime = null, // TODO consider defaulting a fixed number of hours as adjudications have no concept of end time.
+    priority = priorities?.let { pList -> getPriority(null, pList) } ?: EventType.ADJUDICATION_HEARING.defaultPriority,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:132]: Yes - as discussed - I think 2 hours beyond the start time. Just needs a block of time to appear on schedules.

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:132]: For now a hard-coded value. Later we could configure if other prisons need it.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+fun OffenderAdjudicationHearing.toScheduledEvent(
+  prisonCode: String,
+  priorities: List<Priority>?,
+) = let {
+  ScheduledEvent(
+    prisonCode = prisonCode,
+    eventId = it.hearingId,
+    bookingId = null,
+    locationId = it.internalLocationId,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:119]: Do we have the booking ID available to us at this point to populate?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    location = it.internalLocationDescription,
+    eventClass = null,
+    eventStatus = it.eventStatus,
+    eventType = EventType.ADJUDICATION_HEARING.name,
+    eventTypeDesc = null,
+    event = null,
+    eventDesc = null,
+    details = null,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:127]: For `eventDesc` and  `details` - Are there any suitable descriptive things about these hearings without sharing to much on the unlock or calendar?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+  } else {
+    AppointmentLocationSummary(this.locationId, this.agencyId, this.userDescription ?: this.description)
+  }
+
+fun UserDetail?.toSummary(username: String) =
+  if (this == null) {
+    UserSummary(-1, username, "UNKNOWN", "UNKNOWN")
+  } else {
+    UserSummary(this.staffId, this.username, this.firstName, this.lastName)
+  }
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:207]: I like the separation of transform functions!

**Review by timharrison-moj — DISMISSED** [review]
Excellent stuff. Like it. LGTM.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt (hunk 9 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:119]: No. At the moment would need another Prison API call to get them.

We could go back and add the booking ID to the prison API adjudications call I added?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+    } else {
+      when (next.eventCategory) {
+        EventCategory.EDUCATION -> if (category?.startsWith("EDU") == true) listOf(next) else acc
+        EventCategory.GYM_SPORTS_FITNESS -> if (category?.startsWith("GYM") == true) listOf(next) else acc
+        EventCategory.INDUCTION -> if (category == "IND" || category == "INDUC") listOf(next) else acc
+        EventCategory.INDUSTRIES -> if (category == "LACO") listOf(next) else acc
+        EventCategory.INTERVENTIONS -> if (category == "INTERV") listOf(next) else acc
+        EventCategory.LEISURE_SOCIAL -> if (category == "LEI") listOf(next) else acc
+        EventCategory.SERVICES -> if (category == "SERV") listOf(next) else acc
+        else -> {
```
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:179]: I really think this needs addressing. I will look into this as part of this ticket but will pull out into a separate task.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt (hunk 2 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:115]: Looks ok to me as a first iteration.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt (hunk 9 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:119]: It's ok, lets now do it now then. I think some of the others don't have it either, and the UI can assemble this from prisoner-offender-search at the time of displaying, so not to worry.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt (hunk 10 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:129]: Not much else really.  Though I was missing one thing, so added the eventTypeDesc.  Or should this be in the eventDesc?

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/util/PrisonApiTransformations.kt:129]: In theory it shouldn't matter, as I think the unlock display form shows different fields for different event types. Still to be refined what shows for each type though. I was aiming to get `eventType` or `eventTypeDesc` to driving a type of event type display tag, and `eventDescription` to add more detail, and then the `details` to include things like comments or notes, if we get that far.  However, I'll probably have a sweep through to get these more consistent as a separate ticket.

**Review by timharrison-moj — APPROVED** [review]
LGTM. These changes are all contributing to making this area more easily maintainable / changeable, so all good.

**Pull request merged**